docs/String algorithm/AHO-corasick.md

```diff
@@ -0,0 +1,258 @@
+            int curr_node = q.front();
+            q.pop();
+
+            for (auto& [ch, next_node] : trie[curr_node].children) {
+                int fail = trie[curr_node].fail;
+
+                while (fail != 0 && !trie[fail].children.count(ch)) {
+                    fail = trie[fail].fail;
+                }
+
+                if (trie[fail].children.count(ch)) {
+                    trie[next_node].fail = trie[fail].children[ch];
+                } else {
+                    trie[next_node].fail = 0;
+                }
+
+                trie[next_node].output.insert(trie[next_node].output.end(),
+                                              trie[trie[next_node].fail].output.begin(),
+                                              trie[trie[next_node].fail].output.end());
+                q.push(next_node);
+            }
+        }
+    }
+
+    vector<pair<int, string>> search(const string& text) {
+        int node = 0;
+        vector<pair<int, string>> results;
+
+        for (int i = 0; i < text.size(); ++i) {
+            char ch = text[i];
+
+            while (node != 0 && !trie[node].children.count(ch)) {
+                node = trie[node].fail;
+            }
+
+            if (trie[node].children.count(ch)) {
+                node = trie[node].children[ch];
+            } else {
+                node = 0;
+            }
+
+            for (int pattern_idx : trie[node].output) {
+                results.emplace_back(i - patterns[pattern_idx].size() + 1, patterns[pattern_idx]);
+            }
+        }
+
+        return results;
+    }
+};
+
+int main() {
+    vector<string> patterns = {"he", "she", "his", "hers"};
+    string text = "ushers";
+
+    AhoCorasick ac(patterns);
+    vector<pair<int, string>> matches = ac.search(text);
+
+    cout << "Matches found:\n";
+    for (const auto& [idx, pattern] : matches) {
+        cout << "Pattern \"" << pattern << "\" found at index " << idx << endl;
+    }
+
+    return 0;
+}
+```
+
+## Advantages
+
+- **Multiple Pattern Search:** Efficiently searches for multiple patterns in a single pass through the text.
+- **Linear Time Complexity:** After preprocessing the patterns, the search runs in linear time relative to the length of the text.
+- **Failure Links:** These allow the algorithm to efficiently backtrack and reuse previously computed results when mismatches occur.
+
+## Applications
+
+- **Text Search:** Useful in applications such as search engines, text editors, and data mining where multiple patterns need to be searched.
+- **Intrusion Detection Systems:** Often used to detect multiple predefined patterns in network traffic.
+- **DNA Sequence Analysis:** Helps in finding multiple patterns within genetic sequences.
+
+## Conclusion
+
+The **Aho-Corasick Algorithm** is an optimal solution for problems involving multiple pattern matching. Its combination of a trie and failure links ensures that it can handle large datasets and multiple queries efficiently.
+```
+
+---
```

Comment:
delete line 256 to 258  we not need this

Reply:
done